module/zfs/dsl_scan.c

```diff
@@ -50,30 +50,114 @@
 #include <sys/sa_impl.h>
 #include <sys/zfeature.h>
 #include <sys/abd.h>
+#include <sys/range_tree.h>
 #ifdef _KERNEL
 #include <sys/zfs_vfsops.h>
 #endif
 
+/*
+ * Grand theory statement on scan queue sorting
+ *
+ * Scanning is implemented by recursively traversing all indirection levels
+ * in an object and reading all blocks referenced from said objects. This
+ * results in us approximately traversing the object from lowest logical
+ * offset to the highest. Naturally, if we were simply read all blocks in
+ * this order, we would require that the blocks be also physically arranged
+ * in sort of a linear fashion on the vdevs. However, this is frequently
+ * not the case on pools. So we instead stick the I/Os into a reordering
+ * queue and issue them out of logical order and in a way that most benefits
+ * physical disks (LBA-order).
+ *
+ * Queue management:
+ *
+ * Ideally, we would want to scan all metadata and queue up all leaf block
+ * I/O prior to starting to issue it, because that allows us to do an
+ * optimal sorting job. This can however consume large amounts of memory.
+ * Therefore we continuously monitor the size of the queues and constrain
+ * them to 5% (zfs_scan_mem_lim_fact) of physmem. If the queues grow larger
+ * than this limit, we clear out a few of the largest extents at the head
+ * of the queues to make room for more scanning. Hopefully, these extents
+ * will be fairly large and contiguous, allowing us to approach sequential
+ * I/O throughput even without a fully sorted tree.
+ *
+ * Metadata scanning takes place in dsl_scan_visit(), which is called from
+ * dsl_scan_sync() every spa_sync(). If we have either fully scanned all
+ * metadata on the pool, or we need to make room in memory because our
+ * queues are too large, dsl_scan_visit() is postponed and
+ * scan_io_queues_run() is called from dsl_scan_sync() instead. That means,
+ * metadata scanning and queued I/O issuing are mutually exclusive. This is
+ * to provide maximum sequential I/O throughput for the queued I/O issue
+ * process. Sequential I/O performance is significantly negatively impacted
+ * if it is interleaved with random I/O.
+ *
+ * Backwards compatibility
+ *
+ * This new algorithm is backwards compatible with the legacy on-disk data
+ * structures. If imported on a machine without the new sorting algorithm,
+ * the scan simply resumes from the last checkpoint.
```

Comment:
Only gang blocks are scrubbed immediately. I ran some performance tests trying to scrub indirect blocks inline (as long as we were reading them from disk anyway) and it actually made performance about 8% worse.